tensorflow/core/kernels/data/prefetch_dataset_op.cc

```diff
@@ -112,20 +113,14 @@ class PrefetchDatasetOp::Dataset : public DatasetBase {
       slack_us_ = 0;
     }
 
-    ~Iterator() override {
-      // Signal the prefetch thread to terminate it. We will then
-      // join that thread when we delete `this->prefetch_thread_`.
-      //
-      // TODO(mrry): Replace this cancellation logic with a
-      // CancellationManager. The syntax would be more heavyweight,
-      // but it would be possible to thread a cancellation manager
-      // through the IteratorContext to upstream,
-      // potentially-blocking iterators, when we add these.
-      {
-        mutex_lock l(mu_);
-        cancelled_ = true;
-        cond_var_.notify_all();
-      }
+    ~Iterator() override { deregister_fn_(); }
+
+    // Signal the prefetch thread to terminate it. We will then join that
+    // thread when we delete `this->prefetch_thread_`.
+    void Cancel() {
+      mutex_lock l(mu_);
+      cancelled_ = true;
```

Comment:
the whole point of using the cancellation manager is to get rid the `cancelled_` bit ... the runner thread should instead be using `cancellation_manager_.IsCancelled()` and the destructor should be doing cancellation ... you should not need this `Cancel` method.

Reply:
Gotcha!